README.md

|Desktop View| Mobile View |
|--|--|
| ![desktop view](https://github.com/SirusCodes/mobile_web_view/blob/master/screenshots/desktop_view.png?raw=true) | ![mobile view](https://github.com/SirusCodes/mobile_web_view/blob/master/screenshots/mobile_view.png?raw=true) |


# Mobile web view
![](https://img.shields.io/github/license/SirusCodes/mobile_web_view) ![](https://img.shields.io/pub/v/mobile_web_view)

Wanted to show your awesome mobile app to other without letting them into the hassle of installing it? Thanks to flutter web (currently in beta) we have covered you. 

## Installation
**Make sure that you are on beta and web is enabled if not then, [here](https://flutter.dev/docs/get-started/web)**

Add in your pubspec.yaml

	dependencies:
	 mobile_web_view: ^0.1.2

Then import it in your main

	import 'package: mobile_web_view/mobile_web_view.dart';

Warp MobileWebView to your initial route

 

    home: MobileWebView(
            statusBarIconColor: Colors.white,
            content: Text("MobileWebView")
            child: MyHomePage(title: 'Flutter Demo Home Page'),
          ),

|Prams| Description |
|--|--|
| statusBarIconColor | As it says it defines the color of status bar icons and clock in desktop view (default is black)   |
| content | It is to define which widgets to show on the right side of the screen in desktop view (default is SizedBox.expand()) |
| backgroundColor | It sets background color for desktop webview (default is color of Material.canvas) |

## Want to contribute?
A help is always welcomed, check our [CONTRIBUTING.md](https://github.com/SirusCodes/mobile_web_view/blob/master/CONTRIBUTING.md) and don't forget to add yourself to [CONTRIBUTORS.md](https://github.com/SirusCodes/mobile_web_view/blob/master/CONTRIBUTORS.md) 


CHANGELOG.md
## [0.2.0-nullsafety.0] - 23/11/2020

Null-Safety

## [0.1.0] - 20/05/2020

statusBarIconColor, content

## [0.1.1] - 20/05/2020

added documentation comments for the widgets

## [0.1.2] - 03/09/2020

Added background color attribute

lib/mobile_web_view.dart
library mobile_web_view;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

import 'src/mobile_frame.dart';

class MobileWebView extends StatelessWidget {
  const MobileWebView({
    Key? key,
    required this.child,
    this.content = const SizedBox.expand(),
    this.statusBarIconColor = Colors.black87,
    this.backgroundColor,
  }) : super(key: key);

  /// It is your main app that will show up under the mobile frame
  /// and on mobile screen(if on mobile)
  final Widget child;

  /// Content is the widget that is show on the right side
  ///
  /// [default is SizedBox.expand()]
  final Widget content;

  /// This will set the color for status icons
  ///
  /// [default is Colors.black87]
  final Color statusBarIconColor;

  /// This will set the background color
  ///
  /// [default is color of MaterialType.canvas]
  final Color? backgroundColor;
  @override
  Widget build(BuildContext context) {
    if (kIsWeb && MediaQuery.of(context).size.width > 600)
      return Material(
        color: backgroundColor,
        child: Row(
          children: <Widget>[
            Expanded(
              child: Center(
                child: FittedBox(
                  child: MobileFrame(
                    child: child,
                    iconColor: statusBarIconColor,
                  ),
                ),
              ),
            ),
            Expanded(child: content),
          ],
        ),
      );
    return child;
  }
}


flutter_service_worker.js
'use strict';
const MANIFEST = 'flutter-app-manifest';
const TEMP = 'flutter-temp-cache';
const CACHE_NAME = 'flutter-app-cache';
const RESOURCES = {
  "assets/AssetManifest.json": "6d892bd7968a2f8fd7f83aca0676ff94",
"assets/assets/github.png": "ef7a02b69836dc8b6a732a54c4200dcb",
"assets/FontManifest.json": "7b2a36307916a9721811788013e65289",
"assets/fonts/MaterialIcons-Regular.otf": "a68d2a28c526b3b070aefca4bac93d25",
"assets/NOTICES": "0adc0a64b6df2a50b92999223117f256",
"favicon.png": "5dcef449791fa27946b3d35ad8803796",
"icons/Icon-192.png": "ac9a721a12bbc803b44f645561ecb1e1",
"icons/Icon-512.png": "96e752610906ba2a93c65f8abe1645f1",
"index.html": "55efb59cfa049b99c8d71cd3ca524fbc",
"/": "55efb59cfa049b99c8d71cd3ca524fbc",
"main.dart.js": "0f02b1bd7d65e6c9219900913006ae84",
"manifest.json": "9e7b34fd7c291ca2523123ef20497930"
};

// The application shell files that are downloaded before a service worker can
// start.
const CORE = [
  "/",
"main.dart.js",
"index.html",
"assets/NOTICES",
"assets/AssetManifest.json",
"assets/FontManifest.json"];

// During install, the TEMP cache is populated with the application shell files.
self.addEventListener("install", (event) => {
  return event.waitUntil(
    caches.open(TEMP).then((cache) => {
      // Provide a 'reload' param to ensure the latest version is downloaded.
      return cache.addAll(CORE.map((value) => new Request(value, {'cache': 'reload'})));
    })
  );
});

// During activate, the cache is populated with the temp files downloaded in
// install. If this service worker is upgrading from one with a saved
// MANIFEST, then use this to retain unchanged resource files.
self.addEventListener("activate", function(event) {
  return event.waitUntil(async function() {
    try {
      var contentCache = await caches.open(CACHE_NAME);
      var tempCache = await caches.open(TEMP);
      var manifestCache = await caches.open(MANIFEST);
      var manifest = await manifestCache.match('manifest');

      // When there is no prior manifest, clear the entire cache.
      if (!manifest) {
        await caches.delete(CACHE_NAME);
        contentCache = await caches.open(CACHE_NAME);
        for (var request of await tempCache.keys()) {
          var response = await tempCache.match(request);
          await contentCache.put(request, response);
        }
        await caches.delete(TEMP);
        // Save the manifest to make future upgrades efficient.
        await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
        return;
      }

      var oldManifest = await manifest.json();
      var origin = self.location.origin;
      for (var request of await contentCache.keys()) {
        var key = request.url.substring(origin.length + 1);
        if (key == "") {
          key = "/";
        }
        // If a resource from the old manifest is not in the new cache, or if
        // the MD5 sum has changed, delete it. Otherwise the resource is left
        // in the cache and can be reused by the new service worker.
        if (!RESOURCES[key] || RESOURCES[key] != oldManifest[key]) {
          await contentCache.delete(request);
        }
      }
      // Populate the cache with the app shell TEMP files, potentially overwriting
      // cache files preserved above.
      for (var request of await tempCache.keys()) {
        var response = await tempCache.match(request);
        await contentCache.put(request, response);
      }
      await caches.delete(TEMP);
      // Save the manifest to make future upgrades efficient.
      await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
      return;
    } catch (err) {
      // On an unhandled exception the state of the cache cannot be guaranteed.
      console.error('Failed to upgrade service worker: ' + err);
      await caches.delete(CACHE_NAME);
      await caches.delete(TEMP);
      await caches.delete(MANIFEST);
    }
  }());
});

// The fetch handler redirects requests for RESOURCE files to the service
// worker cache.
self.addEventListener("fetch", (event) => {
  var origin = self.location.origin;
  var key = event.request.url.substring(origin.length + 1);
  // Redirect URLs to the index.html
  if (event.request.url == origin || event.request.url.startsWith(origin + '/#')) {
    key = '/';
  }
  // If the URL is not the RESOURCE list, skip the cache.
  if (!RESOURCES[key]) {
    return event.respondWith(fetch(event.request));
  }
  event.respondWith(caches.open(CACHE_NAME)
    .then((cache) =>  {
      return cache.match(event.request).then((response) => {
        // Either respond with the cached resource, or perform a fetch and
        // lazily populate the cache. Ensure the resources are not cached
        // by the browser for longer than the service worker expects.
        var modifiedRequest = new Request(event.request, {'cache': 'reload'});
        return response || fetch(modifiedRequest).then((response) => {
          cache.put(event.request, response.clone());
          return response;
        });
      })
    })
  );
});

self.addEventListener('message', (event) => {
  // SkipWaiting can be used to immediately activate a waiting service worker.
  // This will also require a page refresh triggered by the main worker.
  if (event.data === 'skipWaiting') {
    return self.skipWaiting();
  }

  if (event.message === 'downloadOffline') {
    downloadOffline();
  }
});

// Download offline will check the RESOURCES for all files not in the cache
// and populate them.
async function downloadOffline() {
  var resources = [];
  var contentCache = await caches.open(CACHE_NAME);
  var currentContent = {};
  for (var request of await contentCache.keys()) {
    var key = request.url.substring(origin.length + 1);
    if (key == "") {
      key = "/";
    }
    currentContent[key] = true;
  }
  for (var resourceKey in Object.keys(RESOURCES)) {
    if (!currentContent[resourceKey]) {
      resources.push(resourceKey);
    }
  }
  return contentCache.addAll(resources);
}


lib/src/mobile_frame.dart
import 'package:flutter/material.dart';

class MobileFrame extends StatelessWidget {
  const MobileFrame({
    Key? key,
    required this.child,
    required this.iconColor,
  }) : super(key: key);
  final Widget child;
  final Color iconColor;

  @override
  Widget build(BuildContext context) {
    final mediaQuery = MediaQueryData(
      size: Size(69, 147),
      padding: EdgeInsets.only(
        top: 18,
      ),
      devicePixelRatio: 2,
    );

    return MediaQuery(
      data: mediaQuery,
      child: Padding(
        padding: const EdgeInsets.all(80),
        child: SizedBox.fromSize(
          size: mediaQuery.size * 5,
          child: Container(
            decoration: BoxDecoration(
              color: Colors.black87,
              borderRadius: BorderRadius.circular(40),
              border: Border.all(color: Colors.black87, width: 15),
            ),
            child: ClipRRect(
              borderRadius: BorderRadius.circular(30),
              child: Stack(
                children: [
                  child,
                  Positioned(
                    child: StatusBar(
                      iconColor: iconColor,
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

class StatusBar extends StatelessWidget {
  const StatusBar({Key? key, required this.iconColor}) : super(key: key);
  final Color iconColor;
  @override
  Widget build(BuildContext context) {
    final _time = DateTime.now();
    return Material(
      color: Colors.transparent,
      child: Container(
        height: 30,
        child: Row(
          children: <Widget>[
            SizedBox(width: 12),
            Text(
              "${_time.hour}:${_time.minute}",
              style: TextStyle(
                fontSize: 12,
                fontWeight: FontWeight.bold,
                color: iconColor,
              ),
            ),
            Spacer(),
            Icon(Icons.network_cell, size: 15, color: iconColor),
            Icon(Icons.battery_full, size: 15, color: iconColor),
            SizedBox(width: 12),
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:mobile_web_view/mobile_web_view.dart';
import 'package:url_launcher/url_launcher.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      //
      // Wrap your home widget with MobileWebView
      //
      home: MobileWebView(
        // Sets background color
        backgroundColor: Colors.blueGrey,
        // Sets status bar Icon Color
        statusBarIconColor: Colors.white,
        // Content will build widgets to right side
        content: Content(),
        // Home Widget
        child: MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class Content extends StatelessWidget {
  const Content({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
      children: <Widget>[
        Column(
          children: <Widget>[
            Text(
              "Mobile web view",
              style: TextStyle(
                fontSize: 100,
                fontWeight: FontWeight.bold,
                color: Colors.white,
              ),
              textAlign: TextAlign.center,
            ),
            Text(
              "Null-Safety✅",
              style: TextStyle(
                fontSize: 30,
                fontWeight: FontWeight.bold,
                color: const Color(0xFF16C60C),
              ),
              textAlign: TextAlign.center,
            ),
          ],
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            TextButton(
              onPressed: () =>
                  launch("https://pub.dev/packages/mobile_web_view"),
              child: Text(
                "pub.dev",
                style: TextStyle(
                  fontSize: 40,
                  fontWeight: FontWeight.bold,
                  color: Colors.white,
                  decoration: TextDecoration.underline,
                ),
              ),
            ),
            TextButton(
              onPressed: () =>
                  launch("https://github.com/SirusCodes/mobile_web_view"),
              child: Image.asset(
                "assets/github.png",
                height: 60,
              ),
            ),
          ],
        )
      ],
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
              textAlign: TextAlign.center,
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
